lib/agent.js

```diff
@@ -0,0 +1,70 @@
+'use strict'
+
+const { Pool } = require('../')
+
+const kAgentPoolOpts = Symbol('agent pool opts')
+const kAgentCache = Symbol('agent cache')
+const kAgentRegistry = Symbol('agent registry')
+class Agent {
+    constructor (opts) {
+        this[kAgentPoolOpts] = opts
+        this[kAgentCache] = new Map()
+        this[kAgentRegistry] = new FinalizationRegistry(key => {
+            // get the Pool WeakRef from the cache
+            const ref = this[kAgentCache].get(key)
+            // if the WeakRef exists and the object has been reclaimed
+            // according to mdn the second logic operation here is superfluous as a deref call in a FR will always return undefined
+            if (ref !== undefined && ref.deref() === undefined) {
+                // remove the WeakRef from the cache
+                this[kAgentCache].delete(key)
+            }
+        })
+    }
+
+    get (origin) {
+        // check the cache for an existing Pool
```

Comment:
Later we might need to support 'origin' as an object containing more info to make the key from, in particular TLS stuff.